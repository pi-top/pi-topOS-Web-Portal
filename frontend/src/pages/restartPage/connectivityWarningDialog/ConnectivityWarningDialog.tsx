import React from "react";

import Dialog from "../../../components/atoms/dialog/Dialog";
import Button from "../../../components/atoms/button/Button";

import styles from "./ConnectivityWarningDialog.module.css";

export type Props = {
  active: boolean;
  piTopIpAddress: string;
  shouldMoveAwayFromAp: boolean;
  onSkip: () => void;
  onContinue: () => void;
};

export const getMessage = (shouldMoveAwayFromAp: boolean) => {
  return (
    <>
      <span className={styles.dialogTitle}>{shouldMoveAwayFromAp ? "Reconnect to your Wi-Fi" : "Reconnect to the pi-top network"} </span>
    </>
  );
};

export const getContent = (shouldMoveAwayFromAp: boolean) => {
  if (shouldMoveAwayFromAp) {
    return (
      <>
        Your pi-top is now connected to the Wi-Fi network you configured in a previous step, but your computer is still on the 'pi-top-XXXX' network.
        <br></br>
        <br></br>
        To complete onboarding, please switch your computer to the same Wi-Fi network as your pi-top and click refresh below.
      </>
    )
  }
  return (
    <>
      After your pi-top restarts, you will need to reconnect your computer to the 'pi-top-XXXX' Wi-Fi network to finish onboarding.
      <br></br>
      <br></br>
      This page may not update until you have reconnected to the network.
    </>
  )
}

export default ({
  active,
  piTopIpAddress,
  shouldMoveAwayFromAp,
  onSkip,
  onContinue,
}: Props) => {

  const url = "http://" + piTopIpAddress + "/onboarding/reboot";

  return (
    <Dialog active={active} message={getMessage(shouldMoveAwayFromAp)} className={styles.dialog} testId="connectivity-dialog">
      <div className={styles.content}>
        <div className={styles.message}>
          {getContent(shouldMoveAwayFromAp)}
        </div>

        <div className={styles.buttons}>
          {shouldMoveAwayFromAp ?
            <>
            <div className={styles.refreshButtonContainer}>
              <Button className={styles.refreshButton} unstyled onClick={() => window.open(url, "_self")} >
                Refresh
              </Button>
            </div>

            <div className={styles.skipButtonContainer}>
              <Button className={styles.skipButton} unstyled onClick={() => onSkip()}>
                Skip
              </Button>
            </div>
            </>
        : <>
          <div className={styles.continueButtonContainer}>
            <Button className={styles.continueButton} unstyled onClick={() => onContinue()}>
              Continue
            </Button>
          </div>
          </>
        }
        </div>

      </div>
    </Dialog>
  );
};
