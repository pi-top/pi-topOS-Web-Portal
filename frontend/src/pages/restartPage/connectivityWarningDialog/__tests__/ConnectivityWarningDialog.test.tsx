import React, { ReactNode } from "react";
import ReactDom from "react-dom";
import {
  render,
  fireEvent,
  BoundFunction,
  GetByBoundAttribute,
  QueryByText,
  GetByText,
  RenderResult,
  QueryByBoundAttribute,
  wait,
} from "@testing-library/react";

import ConnectivityWarningDialog, { Props, getContent, getMessage } from "../ConnectivityWarningDialog";

const originalCreatePortal = ReactDom.createPortal;

window.open = jest.fn();

describe("ConnectivityWarningDialog", () => {
  let defaultProps: Props;
  let moveAwayFromApDialog: HTMLElement;
  let queryByText: BoundFunction<QueryByText>;
  let getByText: BoundFunction<GetByText>;
  let queryByTestId: BoundFunction<QueryByBoundAttribute>;
  let queryByAltText: BoundFunction<QueryByBoundAttribute>;
  let queryByLabelText: BoundFunction<QueryByBoundAttribute>;
  let getByLabelText: BoundFunction<GetByBoundAttribute>;
  let rerender: RenderResult["rerender"];

  const ipAddress = "1.1.1.1";

  beforeEach(() => {
    defaultProps = {
      active: true,
      piTopIpAddress: ipAddress,
      onSkip: jest.fn(),
      onContinue: jest.fn(),
      shouldMoveAwayFromAp: true,
    };

    ReactDom.createPortal = jest.fn();
    const createPortalMock = ReactDom.createPortal as jest.Mock;
    createPortalMock.mockImplementation((element: ReactNode) => element)(
      ({
        container: moveAwayFromApDialog,
        queryByText,
        queryByTestId,
        queryByAltText,
        queryByLabelText,
        getByLabelText,
        getByText,
        rerender,
      } = render(<ConnectivityWarningDialog {...defaultProps} />))
    );
  });
  afterEach(() => {
    ReactDom.createPortal = originalCreatePortal;
  });

  describe("when user should switch networks", () =>{
    it("renders the correct message", () => {
      const content = getMessage(true);
      expect(content).toMatchSnapshot();
    });

    it("renders the correct content", () => {
      const content = getContent(true);
      expect(content).toMatchSnapshot();
    });

    it("renders Refresh button", () => {
      expect(queryByText("Refresh")).toBeInTheDocument();
    });

    it("renders Skip button", () => {
      expect(queryByText("Skip")).toBeInTheDocument();
    });

    it("doesn't render the Continue button", () => {
      expect(queryByText("Continue")).not.toBeInTheDocument();
    });


    it("onSkip is called when Skip button is pressed", () => {
      fireEvent.click(getByText("Skip"));
      expect(defaultProps.onSkip).toHaveBeenCalled();
    });

    it("redirects to IP when Refresh button is pressed", async () => {
      fireEvent.click(getByText("Refresh"));
      await wait();
      expect(window.open).toHaveBeenNthCalledWith(
        1,
        "http://" + ipAddress + "/onboarding/reboot",
        "_self"
      )
    });
  })

  describe("when user only connection mode is through AP", () =>{
    beforeEach(()=>{
      defaultProps = {
        ...defaultProps,
        shouldMoveAwayFromAp: false,
      };

      rerender(<ConnectivityWarningDialog {...defaultProps} />);
    })

    it("renders the correct message", () => {
      const content = getMessage(false);
      expect(content).toMatchSnapshot();
    });

    it("renders the correct content", () => {
      const content = getContent(false);
      expect(content).toMatchSnapshot();
    });

    it("renders Continue button", () => {
      expect(queryByText("Continue")).toBeInTheDocument();
    });

    it("doesn't render the Skip button", () => {
      expect(queryByText("Skip")).not.toBeInTheDocument();
    });

    it("doesn't render the Skip button", () => {
      expect(queryByText("Skip")).not.toBeInTheDocument();
    });

    it("onContinue is called when Continue button is pressed", () => {
      fireEvent.click(getByText("Continue"));
      expect(defaultProps.onContinue).toHaveBeenCalled();
    });
  })

});
