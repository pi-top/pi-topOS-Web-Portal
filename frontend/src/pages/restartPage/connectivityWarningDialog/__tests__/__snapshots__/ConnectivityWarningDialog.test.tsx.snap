// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConnectivityWarningDialog when user only connection mode is through AP renders the correct content 1`] = `
<React.Fragment>
  After your pi-top restarts, you will need to reconnect your computer to the 'pi-top-XXXX' Wi-Fi network to finish onboarding.
  <br />
  <br />
  This page may not update until you have reconnected to the network.
</React.Fragment>
`;

exports[`ConnectivityWarningDialog when user only connection mode is through AP renders the correct message 1`] = `
<React.Fragment>
  <span
    className="dialogTitle"
  >
    Reconnect to the pi-top network
     
  </span>
</React.Fragment>
`;

exports[`ConnectivityWarningDialog when user should switch networks renders the correct content 1`] = `
<React.Fragment>
  You've connected your pi-top to Wi-Fi now, but your computer's still on the pi-top's hotspot.
  <br />
  <br />
  Please switch this computer to the Wi-Fi you chose for your pi-top and click refresh below.
</React.Fragment>
`;

exports[`ConnectivityWarningDialog when user should switch networks renders the correct message 1`] = `
<React.Fragment>
  <span
    className="dialogTitle"
  >
    Reconnect to your Wi-Fi
     
  </span>
</React.Fragment>
`;
